internal/services/controller/controller.go

```diff
@@ -211,10 +211,11 @@ func (c *Controller) Run(ctx context.Context) error {
 			const maxItems = 5
 			queueContent := c.debugQueueContent(maxItems)
 			log := c.log.WithField("queue_content", queueContent)
-			// Crash agent in case it's not able to collect full snapshot from informers cache.
-			// TODO (CO-1632): refactor crashing to "normal" exit or healthz metric; abruptly
-			//  stopping the agent does not give it a chance to release leader lock.
-			log.Fatalf("error while collecting initial snapshot: %v", err)
+			log.Errorf("error while collecting initial snapshot: %v", err)
+			c.log.Infof("restarting agent after failure to collect initial snapshot")
+			controllerRuntimeError = err
+			c.triggerRestart()
```

Comment:
we want to terminate the process to keep behavior compatibility. this is only about quitting gracefully/flushing logs etc.